### PR TITLE
:bug: Potential fix to bug causing glyphs to replace others.

### PR DESCRIPTION
**Pull request by Nova-2119**
### Benefits

Hopefully this will stop a bug causing labels to be rendered incorrectly.

### Possible Drawbacks

May not actually fix the issue as I can-not reliably test it as it is an infrequent multi-threading issue for an individual that common enough that a group using constellation frequently will notice it.

### Applicable Issues

#962

### CODE REVIEW

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/GlyphRectangleBuffer.java [hunk 1]
```diff
@@ -256,11 +256,12 @@ int addRectImage(final BufferedImage img, final int extra) {
         //
         final int hashCode = Arrays.hashCode(img.getRGB(0, 0, w, h, null, 0, w));
 
-        Integer rectIndex = memory.putIfAbsent(hashCode, memory.size());
+        
+        Integer rectIndex = memory.get(hashCode);
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/GlyphRectangleBuffer.java:260]: change to primitive `int`

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/GlyphRectangleBuffer.java [hunk 2]
```diff
@@ -356,4 +357,14 @@ private boolean bufferedImagesEqual(BufferedImage img1, BufferedImage img2) {
         }
         return true;
     }
+
+    private synchronized Integer synchronizedAddHashcode(int hashCode, BufferedImage img, int extra, int w, int h) {
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/GlyphRectangleBuffer.java:361]: return type can be primitive `int` and all parameters can be final

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/GlyphRectangleBuffer.java [hunk 3]
```diff
@@ -356,4 +357,14 @@ private boolean bufferedImagesEqual(BufferedImage img1, BufferedImage img2) {
...
+        int value = memory.size();
+        Integer rectIndex = memory.putIfAbsent(hashCode, value);
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/GlyphRectangleBuffer.java:363]: primitive `int`

**Review by antares1470 — CHANGES_REQUESTED** [review]

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/GlyphRectangleBuffer.java (hunk 1 above)
Comment (Nova-2119) [reply CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/GlyphRectangleBuffer.java:260]: Will it not error if memory.get() returns a null?

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/GlyphRectangleBuffer.java (hunk 2 above)
Comment (Nova-2119) [reply CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/GlyphRectangleBuffer.java:361]: Done

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/GlyphRectangleBuffer.java (hunk 3 above)
Comment (Nova-2119) [reply CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/GlyphRectangleBuffer.java:363]: What happens if it returns a null?

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/GlyphRectangleBuffer.java (hunk 1 above)
Comment (antares1470) [reply CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/GlyphRectangleBuffer.java:260]: Right. If the return value can be null then never mind

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/GlyphRectangleBuffer.java (hunk 3 above)
Comment (antares1470) [reply CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/GlyphRectangleBuffer.java:363]: Right. If the return value can be null then never mind

**Review by antares1470 — APPROVED** [review]

**Pull request merged**